tests/apollo/test_skvbc_pyclient.py

```diff
@@ -0,0 +1,191 @@
+                "current primary before sending any requests.")
+        await client.write(protocol.write_req([], [(key, value)], 0))
+        self.assertIsNotNone(client.primary, "A BFT Client reported not " \
+                "knowing the current primary after making a write request.")
+        
+        read_result = await client.read(protocol.read_req([key]))
+        expected_messages_sent = client.msgs_sent
+        self.assertIsNotNone(client.primary, "A BFT Client reported not " \
+                "knowing the current primary after making a read request.")
+        value_read = (protocol.parse_reply(read_result))[key]
+        self.assertEqual(value, value_read, "A BFT Client failed to read a " \
+                "key-value pair from a SimpleKVBC cluster matching the " \
+                "key-value pair it wrote immediately prior to the read.")
+
+        value = protocol.random_value()
+        await client.write(protocol.write_req([], [(key, value)], 1))
+        expected_messages_sent += 1 # We expect this request will only be sent
+                                    # to the primary.
+        self.assertIsNotNone(client.primary, "A BFT Client reported not " \
+                "knowing the current primary after making a write request.")
+        self.assertEqual(expected_messages_sent, client.msgs_sent, "A BFT " \
+                "Client reported having sent an unexpected number of " \
+                "messages after a read request.")
+        
+        read_result = await client.read(protocol.read_req([key]))
+        self.assertIsNotNone(client.primary, "A BFT Client reported not " \
+                "knowing the current primary after making a read request.")
+        value_read = (protocol.parse_reply(read_result))[key]
+        self.assertEqual(value, value_read, "A BFT Client failed to read a " \
+                "key-value pair from a SimpleKVBC cluster matching the " \
+                "key-value pair it wrote immediately prior to the read.")
+
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    async def test_m_of_n_quorum(self, bft_network):
+        """
+        Test that pyclient successfully makes a read request when told to wait
+        for an M of N quorum of a particular size.
+        """
+
+        bft_network.start_all_replicas()
+        client = bft_network.random_client()
+        protocol = skvbc.SimpleKVBCProtocol(bft_network)
+        
+        key = protocol.random_key()
+        value = protocol.random_value()
+        kv_pair = [(key, value)]
+        await client.write(protocol.write_req([], kv_pair, 0))
+
+        quorum = MofNQuorum(bft_network.all_replicas(), 1)
+        read_result = await client.read(protocol.read_req([key]), \
+                                        m_of_n_quorum=quorum)
+        value_read = (protocol.parse_reply(read_result))[key]
+        self.assertEqual(value, value_read, "Using an M of N Quorum of M==1, " \
```

Comment:
For reasons I do not currently fully understand, I have found this assertion to be flaky in my testing. Here is an example of the output I saw from its failure:

```
self.assertEqual(value, value_read, "Using an M of N Quorum of " \
AssertionError: b'MZW86OD9IVKP5LCS2NFYQ' != b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\[39 chars]\x00' : Using an M of N Quorum of M==(F + 1), A BFT Client failed to read a key-value pair from a SimpleKVBC cluster matching the key-value pair it wrote immediately prior to the read.
```

It would appear it is somehow getting a blank value from the server, if I interpret this correctly. I do not know why this would be the case, as I have not seen `test_read_written_value` above fail, and `test_m_of_n_quorum` should be basically the same as that test case, except that it passes an `m_of_n_quorum` parameter to the `read` call. I think the only difference this should make is that the `read` call will only get a response from one replica before returning it rather than validating that response with a larger quorum.

Is there a subtlety of using the `m_of_n_quorum` here that I am missing? Is it possible there are Byzantine-faulty (or at least out-of-sync) replicas I should be worried about in the test network, even though this test case has made no effort to destabilize the cluster?

Reply:
Maybe you can try adding a couple second delay after start_all_replicas similar to what is done in test_skvbc_preexecution.py

Reply:
Please don't add sleeps to make tests "pass".

Reply:
What consensus and actions I need to take can I infer from your comments? Is it safe to assume we agree this issue needs to be debugged further and fixed? Also, can I assume neither of you have previous experience with issues of this particular character or speculations as to what could be going wrong here other than that the issue is sensitive to timing?

Reply:
Yes, this may be a real bug and should be fixed. It just shouldn't be fixed in a non-deterministic manner by adding sleeps.

Reply:
Hi @upshaw-alex , I haven't seen this test fail in CI so far (I reran CI on your PR twice, just in case). If the test fails intermittently, we can disable it and investigate separately. For now, let's merge this PR as is. I agree with @andrewjstone that adding a sleep is not the right approach here.